core/ledger/kvledger/txmgmt/privacyenabledstate/namespace_provider.go

```diff
@@ -0,0 +1,139 @@
+/*
+Copyright IBM Corp. All Rights Reserved.
+
+SPDX-License-Identifier: Apache-2.0
+*/
+
+package privacyenabledstate
+
+import (
+	"github.com/hyperledger/fabric-protos-go/ledger/queryresult"
+	commonledger "github.com/hyperledger/fabric/common/ledger"
+	"github.com/hyperledger/fabric/core/ledger"
+	"github.com/hyperledger/fabric/core/ledger/kvledger/txmgmt/statedb"
+	"github.com/pkg/errors"
+)
+
+// NamespaceProvider implements statedb.NameNamespaceProvider interface
+type NamespaceProvider struct {
+	channelInfoProvider
+	ledger.DeployedChaincodeInfoProvider
+}
+
+// channelInfoProvider interface enables ledger components to retrieve channel config data in all the
+// config blocks (current and previous config blocks).
+type channelInfoProvider interface {
+	// GetAllMSPIDs retrieves the MSPIDs of application organizations in all the channel configurations,
+	// including current and previous channel configurations.
+	GetAllMSPIDs() ([]string, error)
+}
+
+// NewNamespaceProvider creates a new instance of NamespaceProvider
+func NewNamespaceProvider(channelInfoProvider channelInfoProvider, deployedccInfoProvider ledger.DeployedChaincodeInfoProvider) *NamespaceProvider {
+	return &NamespaceProvider{channelInfoProvider, deployedccInfoProvider}
+}
+
+// GetStateDBNamespaces returns the namespaces used by statedb, including:
+// - namespaces for each chaincode and its explicit collections and implicit collections,
+// - special namespaces ("", _lifecycle, lscc)
+// - namespaces for new lifecycle management namespace with implicit collections
+func (p *NamespaceProvider) GetStateDBNamespaces(channelName string, vdb statedb.VersionedDB) ([]string, error) {
+	mspids, err := p.GetAllMSPIDs()
+	if err != nil {
+		return nil, err
+	}
+	implicitCollNames := p.ImplicitCollectionNames(mspids)
+
+	chaincodesInfo, err := p.AllChaincodesInfo(channelName, &simpleQueryExecutor{vdb})
+	if err != nil {
+		return nil, err
+	}
+
+	// iterate each chaincode, add namespaces for chaincodes, implicit collections, and explicit collections
+	retNamespaces := []string{}
+	for _, ccInfo := range chaincodesInfo {
+		ccName := ccInfo.Name
+		retNamespaces = append(retNamespaces, ccName)
+		for _, implicitCollName := range implicitCollNames {
+			retNamespaces = append(retNamespaces, derivePvtDataNs(ccName, implicitCollName))
+			retNamespaces = append(retNamespaces, deriveHashedDataNs(ccName, implicitCollName))
```

Comment:
@cendhu thanks for the comment. Initially I thought it would require new dependency to check peer membership for a collection, but just found that ledger already had MembershipProvider interface. Will add code to check eligibility before adding namespaces for implicit collections.

Reply:
For an implicit collection, this might work. For explicit collections, we only fetch the recent collection config using `DeployedCCInfoProvider.AllChaincodeInfo`. To find all members (including the past members) of a collections, we need to get all collection config from the collection config store. I am not sure whether this complexity is needed. Hence, I am okay if we can add a comment to avoid this confusion (as it is a temporary code).